fiona/ogrext.pyx

```diff
@@ -712,30 +737,46 @@ cdef class Session:
             ignore_fields = set()
 
         cogr_featuredefn = OGR_L_GetLayerDefn(self.cogr_layer)
+
         if cogr_featuredefn == NULL:
             raise ValueError("Null feature definition")
 
         encoding = self._get_internal_encoding()
 
-        n = OGR_FD_GetFieldCount(cogr_featuredefn)
+        num_fields = OGR_FD_GetFieldCount(cogr_featuredefn)
 
-        for i from 0 <= i < n:
+        for i from 0 <= i < num_fields:
             cogr_fielddefn = OGR_FD_GetFieldDefn(cogr_featuredefn, i)
+
             if cogr_fielddefn == NULL:
                 raise ValueError("NULL field definition at index {}".format(i))
 
             key_c = OGR_Fld_GetNameRef(cogr_fielddefn)
+
             if key_c == NULL:
                 raise ValueError("NULL field name reference at index {}".format(i))
+
             key_b = key_c
             key = key_b.decode(encoding)
+
             if not key:
                 warnings.warn("Empty field name at index {}".format(i), FeatureWarning)
 
             if key in ignore_fields:
                 continue
 
+            # See gh-1178 for an example of a pathological collection
+            # with multiple identically name fields.
+            if key in props:
+                log.warning(
+                    "Field name collision detected, field is skipped: i=%r, key=%r",
+                    i,
+                    key
+                )
+                continue
```

Comment:
This preserves existing behavior, field skipping, but adds a warning level log message. I decided against raising a warning, because there's nothing you can do about the fields and schema if you were to handle the warning.